Misc Additions to Local FPGA Branch

abejgonzalez opened the pull request:
**Related issue**: #666 

<!-- choose one -->
**Type of change**: new feature

<!-- choose one -->
**Impact**: rtl change + other

**Release Notes**
This builds off of #666 to have it use the Chipyard make system and try to make the compile work with the newest RC changes. Note, this doesn't check for RTL functionality... still needs to be checked on an Arty board.

colinschmidt commented on fpga/.gitignore:
```diff
@@ -0,0 +1,3 @@
+*
```
I think we should make the builds in here look more standard so we can change this ignore all to something like `generated-src`,`output`, etc.

abejgonzalez replied:
Reasonable goal.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+#########################################################################################
+all_vsrcs := \
+	$(sim_vsrcs) \
+	$(base_dir)/generators/sifive-blocks/vsrc/SRLatch.v \
```
Why these included via the standard firrtl method?

abejgonzalez replied:
These are just black boxes and do not have a resource associated with them. So we would have to make a PR for these to be included through the "normal automatic" way.

abejgonzalez replied:
For now, I just blindly included them, later I think a PR should be opened in both places to make this automatic.

colinschmidt replied:
Issue to track?

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+	$(build_dir)/$(long_name).rom.v
+
+#########################################################################################
+# build rom for the fpga
```
rom->bootrom

abejgonzalez replied:
Sure. Just copied the names.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+export LONG_NAME=$(long_name)
+export ROMCONF=$(build_dir)/$(long_name).rom.conf
+
+romgen := $(build_dir)/$(long_name).rom.v
```
I don't like this name. It is the generated rom not a rom generator. I think `rom_v` `rom_file` etc are better names.

abejgonzalez replied:
Sure. Just copied the names.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+# combine all sources into single .F
+f := $(build_dir)/$(long_name).vsrcs.F
+$(f): $(sim_common_files) $(all_vsrcs)
+	$(foreach file,$(all_vsrcs),echo "$(file)" >> $@;)
```
Does vivado handle duplicates? I think we have had dedupe code for these `*.f` files in the past.

abejgonzalez replied:
The `sim_common_files` is deduped. It shouldn't collide with the `all_vsrcs`. As for Vivado... I don't know.

colinschmidt replied:
Ok cool as long as we reusing that code.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+	$(foreach file,$(all_vsrcs),echo "$(file)" >> $@;)
+	cat $(sim_common_files) >> $@
+
+bit := $(build_dir)/obj/$(MODEL).bit
```
Again I think we should disambiguate the name of the variable for the output from the name of the target.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+bit: $(bit)
+
+# Build .mcs
+mcs := $(build_dir)/obj/$(MODEL).mcs
```
As above. bad name.

colinschmidt commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+#########################################################################################
+# mircosemi rules
+#########################################################################################
+# Build Libero project
```
Do we know what this is? Does it apply to the subset we are including?

abejgonzalez replied:
I don't know, I just copied this. I think we are just using the Vivado stuff.

colinschmidt replied:
I think we should delete it for now, rather than keep around dead code (and perhaps make users think we support this).

colinschmidt commented on fpga/bootrom/xip/Makefile:
```diff
@@ -1,12 +1,17 @@
+$(dtb): $(BUILD_DIR)/$(LONG_NAME).dts
 	dtc -I dts -O dtb -o $@ $<
 
 .PHONY: dtb
```
Hopefully we can apply the same `s/filename[^:]/filename_file/` rewrite rule to all these intermediate files.

colinschmidt commented on fpga/bootrom/xip/Makefile:
```diff
@@ -35,11 +40,11 @@ hex: $(hex)
 
```
I can't comment on the above lines since they aren't in the diff for this PR, but I have one question.
Is the hardcoded target_address coming from an RTL config fragment? If so we should think about how to do this without repeating ourselves.

abejgonzalez replied:
Yea. Im not sure. I need to look into this more... they end up using a `MaskROM` over a `BootROM` in the code which is a bit confusing to me.

colinschmidt replied:
I believe `MaskROM` implies a specific kind of `ROM` that is reprogrammable with just a single VLSI mask update. I think there is an overload used here sometimes for an FPGA `ROM` that can be updated without the rest of the bit stream being changed.

Besides that definition, I think a `MaskROM` can be a `BootROM` (a `ROM` that contains code to boot, and is where the core resets to), but not all `BootROM`s would be `MaskROM`s and not all `MaskROM`s would be `BootROM`s.

abejgonzalez replied:
So it looks like the `MaskROM` is used to load in a ROM after elaboration using `readmemh`. This differs from the BootROM which bakes the `.img` directly into the module in Chisel.

abejgonzalez replied:
Should be good to just swap `MaskROM` for the `BootROM`

colinschmidt commented on fpga/bootrom/xip/Makefile:
```diff
@@ -35,11 +40,11 @@ hex: $(hex)
 
 .PHONY: clean
 clean::
-	rm -rf $(hex) $(elf) 
```
Nice

abejgonzalez replied:
#vim>emacs

colinschmidt commented on fpga/src/main/scala/arty/Platform.scala:
```diff
@@ -51,6 +51,7 @@ class E300ArtyDevKitPlatformIO(implicit val p: Parameters) extends Bundle {
 //-------------------------------------------------------------------------
 
 class E300ArtyDevKitPlatform(implicit val p: Parameters) extends Module {
+  //val sys = Module(LazyModule(new E300ArtyDevKitSystem).module) This can be DigitalTop?
```
Is this just remnants from some other testing?

abejgonzalez replied:
Oh. Yea. This should be deleted.

colinschmidt commented on fpga/src/main/scala/arty/System.scala:
```diff
@@ -31,6 +31,12 @@ class E300ArtyDevKitSystem(implicit p: Parameters) extends RocketSubsystem
+  val maskROMs = p(MaskROMLocated(location)).map { MaskROM.attach(_, this, CBUS) }
+
+  val maskROMResetVectorSourceNode = BundleBridgeSource[UInt]()
+  tileResetVectorNexusNode := maskROMResetVectorSourceNode
```
It seems weird to support both `bootROM` and `maskROM` above but then not do some connection here.
Do we require a `maskROM` but optionally allow a `bootROM` that never can be the reset location?

abejgonzalez replied:
Yea. This was a hack to try and do what the original code wanted. I need to look into this.

colinschmidt commented on generators/utilities/src/main/scala/Simulator.scala:
```diff
@@ -82,26 +85,31 @@ object GenerateSimFiles extends App with HasGenerateSimConfig {
-      "/vsrc/TestDriver.v",
-    )
-  })
+    ) ++ (sim match {
```
Do we need this empty `Seq` at the beginning?

abejgonzalez replied:
Sorry? Are you referring to L90? If so, yes, you want to not include the simulation files (or you have to filter them out in the `Makefile` when using `sim_common_files`)

colinschmidt replied:
Oh this has changed I think since I looked at it before, or I misread before.
I thought it was:
```
Seq(
) ++ (sim match {
...
```

colinschmidt reviewed (approved):
Seem pretty good. I think we don't need to consume the weird style habits in the makefile, and instead can make the names more similar to the rest of chipyard.

Also obviously testing would be good, but I assume that can happen after the merge into the orig PR.

Not blocking merge since this is just a PR to a PR

jerryz123 commented on fpga/Makefile:
```diff
@@ -0,0 +1,124 @@
+GENERATOR_PACKAGE := chipyard
+TB                := none # unused
+TOP               := E300ArtyDevKitPlatform
+
```
Should the goal be to eventually eliminate most of these overrides? I believe eventually everything can just live as part of the core chipyard generator.

colinschmidt replied:
I think thats related to the other checkbox in my review of the original PR. I believe the answer is yes.

abejgonzalez replied:
Yes. The goal is to get rid of these. I just hard-coded them so that this particular example works all the time. Ideally, this will change with more configuration exposed.

abejgonzalez commented:
Updated. Now uses the CY BootRom, addressed some comments, and now uses `DigitalTop` (aka can now access all CY projects). Need to see what the `MockAON` is since I don't think it can be "disabled" but so far this is getting closer.
